### ISSUE TITLE
Diving Deep Into The Golang Channels.

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9284]
Diving Deep Into The Golang Channels.<br>
<br>
https://ift.tt/PGEsQK5<br>
<br>
Ankur Anand<br>
<br>

<div><div><div><div>
<div>
<p></p>
<h1>Diving Deep Into The Golang Channels.</h1>
<p>An “ins and out” of the internal implementation of the Golang channels and its related operations.</p>
<p>Concurrency in Golang is much more than just syntax.</p>
<blockquote><p>It a design pattern.</p></blockquote>
<p>A pattern that is a repeatable solution to a commonly occurring problem while working with concurrency, because even</p>
<blockquote><p>Concurrency Needs to be Synchronized.</p></blockquote>
<p>And Go relies on a concurrency model called CSP ( Communicating Sequential Processes), to achieve this pattern of synchronization through Channel. Go core philosophy for concurrency is</p>
<p><em>Do not communicate by sharing memory; instead, share memory by communicating.</em></p>
<p>But Go also trusts you to do the right thing. So Rest of the post will try to open this envelope of Go philosophy and how channels — using a queue to achieve the same.</p>
<h1>What it takes to be a Channel.</h1>
<pre><span>func goRoutineA(a &lt;-chan int) {<br> val := &lt;-a<br> fmt.Println("goRoutineA received the data", val)<br>}</span><span>func main() {<br> ch := make(chan int)<br> go goRoutineA(ch)<br> time.Sleep(time.Second * 1)<br>}</span></pre>
<figure></figure><figure></figure><p>So it’s Responsibility of channel in Go to make the Goroutine runnable again that is blocked on the channel while receiving the data or sending the data.</p>
<p><em>If you are unfamiliar with Go Scheduler please read this nice introduction about it. </em><a rel="noopener ugc nofollow" href="https://morsmachine.dk/go-scheduler" target="_blank"><em>https://morsmachine.dk/go-scheduler</em></a></p>
<h1>Channel Structure</h1>
<p>In Go, the “channel structure” is the basis of message passing between Goroutine. So What does this channel structure looks like after we create it?</p>
<pre><span>ch := make(chan int, 3)</span></pre>
<figure><figcaption>Buffered Channel Structure after channel creation during runtime.</figcaption></figure><p>Looks good, good. But what does this really mean? and from where channel gets its structure. Let’s look at a few important structs before going any further.</p>
<h2>hchan struct</h2>
<p>When we write <code>make(chan int, 2)</code>channel is created from the hchan struct, which has the following fields.</p>
<figure><figcaption>hchan and waitq structs</figcaption></figure><p>Lets put descriptions to a few fields that we encountered in the channel structure.<br><br><strong>dataqsize</strong> Is the size of the buffer mentioned above, that is make(chan T, N), the N.</p>
<p><strong>elemsize</strong> Is the size of a channel corresponding to a single element.</p>
<p><strong>buf </strong>is the <em>circular queue where our data is actually stored. (used only for buffered channel)</em></p>
<p><strong>closed</strong> Indicates whether the current channel is in the closed state. After a channel is created, this field is set to 0, that is, the channel is open; by calling close to set it to 1, the channel is closed.</p>
<p><strong>sendx</strong> and <strong>recvx</strong> is state field of a ring buffer, which indicates the current index of buffer — backing array from where it can send data and receive data.</p>
<p><strong>recvq</strong> and <strong>sendq</strong> waiting queues, which are used to store the blocked goroutines while trying to read data on the channel or while trying to send data from the channel.</p>
<p><strong>lock</strong> To lock the channel for each read and write operation as sending and receiving must be mutually exclusive operations.</p>
<p>So what is this <strong>sudog</strong>?</p>
<h2>sudog struct</h2>
<blockquote><p>sudog represent the goroutine.</p></blockquote>
<figure><figcaption>Important Field of sudog struct for channel</figcaption></figure><p>Let’s try to wrap our head around the channel structure again step by step. It’s important to have a clear picture of these as this is what gives channel the power in Go.</p>
<figure><figcaption>Example Code</figcaption></figure><p>What will be the structure of the channel before <strong>line No 22?</strong></p>
<figure><figcaption>Chan Struct at the runtime</figcaption></figure><p>Pay attention to highlighted line no 47 and 48 above. Remember <strong>recvq</strong> description from above</p>
<blockquote><p><strong>recvq </strong>are used to store the blocked goroutines which are trying to read data from the channel.</p></blockquote>
<p>In Our Example Code before line 22 there are two goroutines (<code>goroutineA</code> and <code>goroutineB</code> ) trying to read data from the channel <code>ch</code></p>
<p>Since before line 22 on a channel, there is no data we have put on the channel so both the goroutines blocked for receive operation have been wrapped inside the <code>sudog</code> struct and is present on the <strong>recvq </strong>of the channel.</p>
<blockquote><p>sudog represent the goroutine.</p></blockquote>
<p>recvq and sendq are basically linked list, which looks basically as below</p>
<figure><figcaption>Recvq structure</figcaption></figure><p>These structures are really Important,</p>
<p>Let’s see what happens when we try to put the data on the channel <code>ch</code></p>
<h1>Send Opertaion Steps c &lt;- x</h1>
<blockquote><p>Underlying types of send Operations on Channel</p></blockquote>
<ol><li><strong>sending on nil channel</strong></li></ol>
<figure></figure><p>If we are sending on the nil channel the current <code>goroutine</code> will suspend its operation.</p>
<p>2<strong>. sending on the closed channel.</strong></p>
<figure></figure><p>If we try to send data on the closed channel our goroutine panic.</p>
<p>3<strong>. A goroutine is blocked on the channel: the data is sent directly to the goroutine.</strong></p>
<figure></figure><p>This is where <code>recvq</code> structure plays such an important role. If there is any goroutine in the <code>recvq</code> it’s a <em>waiting receiver, </em>and current write operation to channel can directly pass the value to that receiver. Implementation of the send function.</p>
<figure></figure><p>Pay attention to the line number 396 <code>goready(gp, skip+1)</code> the <code>Goroutine</code> which was blocked while waiting for the data has <strong>been made runnable again by calling </strong><code><strong>goready,</strong></code><strong> </strong>and the go scheduler will run the goroutine again.</p>
<p>4<strong>. Buffered Channel if there is currently space available for hchan.buf: put the data in the buffer.</strong></p>
<figure></figure><p><code>chanbuf(c, i)</code> accesses the corresponding memory area.</p>
<p>Determine if hchan.buf has free space by comparing qcount and dataqsiz. <strong>**Enqueue the element by copying the area pointed to by the ep pointer to the ring buffer to send**</strong>, and adjust sendx and qcount.</p>
<p>5. <strong>The hchan.buf is full</strong></p>
<figure></figure><p>Create a <code>goroutine</code> object on the current stack</p>
<p><code>acquireSudog</code> to put the current goroutine in the park state and then add that goroutine in the <code>sendq</code> of the channel.</p>
<h2>Send operation Summary</h2>
<ol>
<li>
<strong>lock</strong> the entire channel structure.</li>
<li>determines writes. Try <code>recvq</code> to take a waiting goroutine from the wait queue, then hand the element to be written directly to the goroutine.</li>
<li>If recvq is Empty, Determine whether the buffer is available. If available, <strong>**copy**</strong> (<code>typedmemmove</code> copies a value of type t to dst from src.`) the data from current goroutine to the buffer.<br><em>_typedmemmove_</em> internally uses <em>memmove — memmove() is used to copy a block of memory from a location to another.</em>
</li>
<li>If the buffer is full then the element to be written is saved in the structure of the currently executing goroutine and the current goroutine is enqueued at <strong>sendq</strong> and suspended, from runtime.</li>
</ol>
<p>Point number 4 is really Interesting.</p>
<blockquote><p>If the buffer is full then the element to be written is saved in the structure of the currently executing goroutine.</p></blockquote>
<p><strong>Read it again, because this is why the unbuffered channel is actually called “unbuffered” even though the “hchan” struct has the “buf” element associated with it</strong>. Because for an unbuffered channel if there is no receiver and if you try to send data, the data will be saved in the <code>elem</code> of the <code>sudog</code> structure. (Holds true for the buffered channel too).</p>
<p>Let me give you an example to clarify the point number 4 in more details. Suppose we have the below code.</p>
<figure><figcaption>Don’t run this code it will cause panic in normal mode.</figcaption></figure><p>What will be the runtime structure of the chan <code>c2</code> at line number10 ?</p>
<figure><figcaption>runtime unbuffered</figcaption></figure><p>You can see even though we have put int value <code>2</code> on the channel the <code>buf</code> does not have the value, but it will be in the <code>sudog</code> structure of the goroutine. As <code>goroutineA</code> tried to send value over to the channel <code>c2</code> and there were no receiver ready, so the <code>goroutineA</code> will be added to <code>sendq</code> list of the channel <code>c2</code> and will be parked as it blocks. We can look into the runtime structure of the blocking <code>sendq</code> to verify.</p>
<figure></figure>
</div>
<div>
<h1>Select</h1>
<blockquote><p>Multiplexing on multiple channel.</p></blockquote>
<figure><figcaption>select channel Example</figcaption></figure><ol><li>Operations are mutually exclusive, so <strong>need to acquire the locks on all involved channels in select case, </strong>which is done by sorting the cases by Hchan address to get the <strong>locking order,</strong> so that it does not lock mutexes of all involved channels at once.</li></ol>
<pre><span>sellock(scases, lockorder)</span></pre>
<p>Each <code>scase</code> in the <code>scases</code> array is a struct which contains the kind of operation in the current case and the channel it’s operating on.</p>
<figure><figcaption>scase</figcaption></figure><p><strong>kind</strong> Is the type of operation in the current case, can be <em>CaseRecv</em>, <em>CaseSend </em>and <em>CaseDefault</em>.</p>
<p>2. Calculate the poll order to shuffle all involved channels to provide the pseudo-random guarantee and traverse all the cases in turn according to the poll order one-by-one to see if any of them is ready for communication. This poll order what makes select operations to not necessarily follow the order declared in the program.</p>
<figure><figcaption>poll order</figcaption></figure><figure><figcaption>cases in select</figcaption></figure><p>3. The select statement can return as long as there is a channel operation that doesn’t block, without even need to touch all channels if the selected one is ready.</p>
<p>3. If no channel currently responds and there is no default statement, current <code>g</code> must currently hang on the corresponding wait queue for all channels according to their case.</p>
<figure><figcaption>park goroutine in select case</figcaption></figure><p><code>sg.isSelect</code> is what indicates that goroutine is participating in the select statement.</p>
<p>4. Receive, Send and Close operation during Select Operation is similar to the general operation of Receive, Send and Close on channels.</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium <br>
<br>
April 28, 2023 at 07:23PM
